// common/utils.js
var npath = require('path')
var fs = require('fs')

var utils = {
    p: function(path) {
        return npath.resolve(path)
    },

    punix: function(path) {
        return utils.p(path).replace('\/', '/')
    },

    ensurePath: function(path) {
        if (!fs.existsSync(path)) {
            utils.ensurePath(npath.dirname(path))
            fs.mkdirSync(path)
        }
    },

    getBuildTime: function() {
        var now = new Date();
        return '' + now.getFullYear() + (now.getMonth() + 1) + now.getDate() + now.getHours() + now.getMinutes();
    },

    /**
     *
     * @param logs: []
     *  $index: string
     *
     *  格式如下
     *
     *  TYPE: message
     *
     *  TYPE 可能是以下的某个值:
     *
     *  error, info, hint, warn
     *
     */
    logs: function(logs) {
        var types = {
            error: function(message) {
                console.error('[ERROR] ' + message)
            },
            info: function(message) {
                console.info('[INFO] ' + message)
            },
            hint: function(message) {
                console.info('[HINT] ' + message)
            },
            warn: function(message) {
                console.warn('[WARNING] ' + message)
            },
            default: function(message) {
                console.log(message)
            }
        }

        logs.map(function(log) {
            var parts = log.split(':')
            var type = parts[0], message = types[type] ? _.trim(parts[1]) : log;
            var logFunc = types[type] || types['default'];
            logFunc(message)
        })
    }
}

module.exports = utils
